File: Python/Algorithms/Searching-Algorithms/interpolation_search.py
# Interpolation search is performed by the formula:
# pos = lo + [ (x-arr[lo])*(hi-lo) / (arr[hi]-arr[Lo]) ]
# where,
# arr[] = Array where elements need to be searched
# x = Element to be searched
# lo = Starting index in arr[]
# hi = Ending index in arr[]

# Function to perform interpolation search
def interpolation_search(array, element):
    lo = 0
    hi = len(array) - 1 

    while lo <= hi and element >= array[lo] and element <= array[hi]:
        # Check if the list is one element list
        if lo == hi:
            if array[lo] == element:
                return lo
            return -1
        
        # Formula for interpolation search
        position = lo + int(((float(hi - lo)/
                    (array[hi]- array[lo])) * (element - array[lo])))

        # Check if the position is correct for the element
        if array[position] == element:
            return position
        
        if array[position] < element:
            lo = position + 1
        
        else:
            hi = position -1
    
    return -1


length = int(input("Enter the number of element you want: "))   # Get the size of the array
my_list = []

# Get the input elements of the list
for i in range(length):
    in_element = int(input(f" Enter {i+1} element: "))
    my_list.append(in_element)

# Get the element whose position is to be known
n= int(input("Enter the element whose position you want to know: "))
index = interpolation_search(my_list, n)

if index != -1:
    print (f"{n} is found at position  ", index + 1) # The output gives the position of the element, not index

else:
    print (f"{n} is not found in the list")

File: C++/Algorithms/Searching-Algorithms/InterpolationSearch.cpp
//Interpolation Search
//The Interpolation Search is an improvement over Binary Search for instances, where the values in a sorted array are uniformly distributed.
// Binary Search always goes to the middle element. On the other hand, interpolation search may go to different locations according to the 
//value of the key being searched.

//Approach: The idea of formula is to return higher value of pos
// when element to be searched is closer to arr[high]. And
// smaller value when closer to arr[lo]
//pos = low + [ (x-arr[low])*(high-low) / (arr[high]-arr[low]) ]

//Time Complexity: O (log (log n)),If elements are uniformly distributed
//Auxiliary Space: O(1)

// C++ implementation of Interpolation Search

#include<bits/stdc++.h> 
using namespace std; 
  

int interpolationSearch(int arr[], int n, int x)        //x is the element to be searched and n is the size of the array
{ 
    // Find indexes of two corners 
    int low = 0, high = (n - 1); 
  
    while (low <= high && x >= arr[low] && x <= arr[high]) 
    { 
        if (low == high) 
        { 
            if (arr[low] == x) return low; 
            return -1; 
        } 
        
        int pos = low + (((double)(high - low) / 
            (arr[high] - arr[low])) * (x - arr[low])); 
  
        if (arr[pos] == x) 
            return pos; 
  
        // If x is larger, x is in upper part 
        if (arr[pos] < x) 
            low = pos + 1; 
  
        // If x is smaller, x is in the lower part 
        else
            high = pos - 1; 
    } 
    return -1; 
} 
  
// Driver Code 
int main() 
{ 
    int arr[] = {10, 12, 13, 16, 18, 19, 20, 21, 
                 22, 23, 24, 33, 35, 42, 47}; 
    int n = sizeof(arr)/sizeof(arr[0]); 
  
    int x = 18; // Element to be searched 
    int index = interpolationSearch(arr, n, x);         //function called
  
    // If element was found 
    if (index != -1) 
        cout << "Element found at index " << index; 
    else
        cout << "Element not found."; 
    return 0; 
} 
  